bridge/tracker/tracker.go

```diff
@@ -0,0 +1,297 @@
+package tracker
+
+import (
+	"context"
+	"encoding/json"
+	"github.com/0xPolygon/polygon-edge/blockchain/storage"
+	"github.com/hashicorp/go-hclog"
+	"github.com/umbracle/go-web3"
+	"math/big"
+)
+
+type Config struct {
+	Confirmations uint64
+	RootchainWS   string
+	DBPath        string
+	ContractABIs  map[string][]string
+}
+
+//	cancellable context for tracker's listening mechanism
+type contextSubscription struct {
+	context context.Context
+	cancel  context.CancelFunc
+}
+
+//	done returns the contexts Done channel.
+func (c *contextSubscription) done() <-chan struct{} {
+	return c.context.Done()
+}
+
+//	Tracker represents an event listener that notifies
+//	for each event emitted by a smart contract on the rootchain.
+//	Events are collected from blocks at the desired depth (confirmations).
+type Tracker struct {
+	logger hclog.Logger
+
+	// required block confirmations
+	confirmations *big.Int
+
+	// event channel
+	eventCh chan []byte
+
+	// cancel subscription
+	ctxSubscription contextSubscription
+
+	//	rootchain subscription object
+	sub subscription
+
+	//	rootchain client
+	client *rootchainClient
+
+	//	db for last processed block number
+	db storage.Storage
+
+	//	events that tracker will listen for
+	contracts []*contractABI
```

Comment:
why not `[]*abi.Event`?

Reply:
I use the `contractABI` struct to keep both the address of the smart contract as well as (abi) events defined in it. Addresses are used in [setupQueryFilter](https://github.com/0xPolygon/polygon-edge/blob/d8ba42ed44ff9eb623c01b2ff90942b5530d7ebb/bridge/tracker/rootchain.go#L82) to narrow down the query. But if topics alone (event signatures) would be enough to match logs, perhaps smart contract addresses are redundant ?

Reply:
No, I thought we could pass parsed event information, not new struct
We need contract address and event